### PR TITLE
Breakpoints for plugins

### PR DESCRIPTION
**Pull request by rageagainsthepc**

### CODE REVIEW

File: vmicore/include/vmicore/plugins/PluginInterface.h [hunk 1]
```diff
@@ -47,6 +49,11 @@ namespace VmiCore::Plugin
 
         virtual void registerShutdownEvent(shutdownCallback_f shutdownCallback) = 0;
 
+        [[nodiscard]] virtual std::shared_ptr<IBreakpoint>
+        createBreakpoint(uint64_t targetVA,
+                         uint64_t processCr3,
```
Comment (thomasdangl) [inline vmicore/include/vmicore/plugins/PluginInterface.h:54]: Can we use some platform agnostic term, e.g., PGD or DTB?

File: vmicore/src/plugins/PluginSystem.cpp [hunk 2]
```diff
@@ -113,6 +115,12 @@ namespace VmiCore
         registeredShutdownCallbacks.push_back(shutdownCallback);
     }
 
+    std::shared_ptr<IBreakpoint> PluginSystem::createBreakpoint(
+        uint64_t targetVA, uint64_t processCr3, const std::function<BpResponse(IInterruptEvent&)>& callbackFunction)
```
Comment (thomasdangl) [inline vmicore/src/plugins/PluginSystem.cpp:119]: Same here

File: vmicore/src/plugins/PluginSystem.h [hunk 3]
```diff
@@ -78,6 +81,11 @@ namespace VmiCore
 
         void registerShutdownEvent(Plugin::shutdownCallback_f shutdownCallback) override;
 
+        std::shared_ptr<IBreakpoint>
+        createBreakpoint(uint64_t targetVA,
+                         uint64_t processCr3,
```
Comment (thomasdangl) [inline vmicore/src/plugins/PluginSystem.h:86]: and here

File: vmicore/src/vmi/InterruptEventSupervisor.cpp [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    void InterruptEventSupervisor::teardown()
+    {
+        clearInterruptEventHandling();
+        singleStepSupervisor->teardown();
+    }
+
+    std::shared_ptr<IBreakpoint> InterruptEventSupervisor::createBreakpoint(
+        uint64_t targetVA, uint64_t processCr3, const std::function<BpResponse(IInterruptEvent&)>& callbackFunction)
```
Comment (thomasdangl) [inline vmicore/src/vmi/InterruptEventSupervisor.cpp:50]: and here

File: vmicore/src/vmi/InterruptEventSupervisor.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        ~InterruptEventSupervisor() noexcept override;
+
+        void initialize() override;
+
+        void teardown() override;
+
+        std::shared_ptr<IBreakpoint>
+        createBreakpoint(uint64_t targetVA,
+                         uint64_t processCr3,
```
Comment (thomasdangl) [inline vmicore/src/vmi/InterruptEventSupervisor.h:54]: : )

File: vmicore/src/vmi/InterruptEventSupervisor.cpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+            vmiInterface->flushV2PCache(LibvmiInterface::flushAllPTs);
+            vmiInterface->flushPageCache();
+            storeOriginalValue(targetPA);
+            enableEvent(targetPA);
+
+            auto targetGFN = targetPA >> PagingDefinitions::numberOfPageIndexBits;
+            // One PageGuard guards a whole memory page on which several interrupts may reside.
+            if (!interruptGuardsByGFN.contains(targetGFN))
+            {
+                interruptGuardsByGFN[targetGFN] = createPageGuard(targetVA, processCr3, targetGFN);
```
Comment (thomasdangl) [inline vmicore/src/vmi/InterruptEventSupervisor.cpp:78]: only for Windows, right?

**Review by thomasdangl — COMMENTED** [review]
Some nits, other than that, looks good

File: vmicore/src/vmi/InterruptEventSupervisor.cpp (hunk 6 above)
Comment (rageagainsthepc) [reply vmicore/src/vmi/InterruptEventSupervisor.cpp:78]: Userspace breakpoints should be hidden on Linux as well, should they not?

**Review by K-Mayer — APPROVED** [review]

File: vmicore/src/vmi/Event.h [hunk 7]
```diff
@@ -1,35 +1,43 @@
...
 #include <cstdint>
 #include <libvmi/events.h>
-#include <memory>
-#include <utility>
+#include <vmicore/vmi/events/IInterruptEvent.h>
 
 namespace VmiCore
 {
-    class Event
+    class Event : public IInterruptEvent
```
Comment (cakeless) [inline vmicore/src/vmi/Event.h:10]: What happends for other types of events ? (e.g. memory / register)

Comment (rageagainsthepc) [reply vmicore/src/vmi/Event.h:10]: Support for those is not implemented yet. We should create a followup PR as soon as we need it.

Comment (cakeless) [inline vmicore/src/vmi/Event.h:10]: Why did you switch the inheritance order ?

**Review by cakeless — APPROVED** [review]

File: vmicore/test/os/windows/SystemEventSupervisor_UnitTest.cpp [hunk 8]
```diff
@@ -21,22 +25,34 @@ namespace VmiCore
...
+                                                                                     activeProcessSupervisor,
+                                                                                     configInterface,
+                                                                                     interruptEventSupervisor,
+                                                                                     logging,
+                                                                                     eventStream);
+        }
     };
 
-    TEST_F(SystemEventSupervisorFixture, teardown_validState_interruptFactoryTeardownCalled)
+    TEST_F(SystemEventSupervisorFixture, teardown_validState_interruptTeardownCalled)
```
[the remark this answers is not in the document]
Comment (rageagainsthepc) [reply vmicore/test/os/windows/SystemEventSupervisor_UnitTest.cpp:48]: Rolls off the tongue quite nicely :P

**Pull request merged**